File: apps/mobile/lib/src/features/auth/domain/usecases/sign_up_usecase.dart
import 'package:equatable/equatable.dart';
import 'package:fpdart/fpdart.dart';

import '../../../../core/error/failures.dart';
import '../../../../core/usecase/usecase.dart';
import '../entities/auth_session.dart';
import '../repositories/auth_repository.dart';

class SignUpParams extends Equatable {
  const SignUpParams({
    required this.email,
    required this.password,
    required this.displayName,
  });
  final String email;
  final String password;
  final String displayName;

  @override
  List<Object?> get props => [email, password, displayName];
}

class SignUpUseCase implements UseCase<AuthSession, SignUpParams> {
  const SignUpUseCase(this._repository);
  final AuthRepository _repository;

  @override
  Future<Either<Failure, AuthSession>> call(SignUpParams params) {
    return _repository.signUp(
      email: params.email,
      password: params.password,
      displayName: params.displayName,
    );
  }
}

File: apps/api/src/features/auth/domain/usecases/sign-up.usecase.ts
import argon2 from 'argon2';
import { AppError } from '@/core/errors/app-error.js';
import type { User } from '../entities/user.js';
import type { AuthRepository } from '../repositories/auth.repository.js';
import type { AuthTokens, TokenService } from '../services/token.service.js';

export interface SignUpInput {
  email: string;
  password: string;
  displayName: string;
}

export interface SignUpResult {
  user: User;
  tokens: AuthTokens;
}

export class SignUpUseCase {
  constructor(
    private readonly authRepository: AuthRepository,
    private readonly tokenService: TokenService,
  ) {}

  async execute(input: SignUpInput): Promise<SignUpResult> {
    const existing = await this.authRepository.findUserByEmail(input.email);
    if (existing) {
      throw AppError.conflict('Email already registered');
    }

    const passwordHash = await argon2.hash(input.password);

    const user = await this.authRepository.createUser({
      email: input.email,
      passwordHash,
      displayName: input.displayName,
    });

    const tokens = await this.tokenService.issueTokens({ sub: user.id, email: user.email });

    return { user, tokens };
  }
}
